runners/flink/src/main/java/org/apache/beam/runners/flink/LyftFlinkPipelineRunner.java

```diff
@@ -0,0 +1,212 @@
+    } catch (Exception e) {
+      throw new RuntimeException(String.format("Job %s failed.", configuration.driverCmd), e);
+    } finally {
+      LOG.info("Stopping job service");
+      runner.stopJobService();
+    }
+    LOG.info("Job submitted successfully.");
+  }
+
+  private static class EntryPointConfiguration {
+    @Option(
+        name = "--driver-cmd",
+        required = true,
+        usage =
+            "Command that launches the Python driver program. "
+                + "(The job service endpoint will be appended as --job_endpoint=localhost:<port>.)")
+    private String driverCmd = null;
+  }
+
+  private static EntryPointConfiguration parseArgs(String[] args) {
+    EntryPointConfiguration configuration = new EntryPointConfiguration();
+    CmdLineParser parser = new CmdLineParser(configuration);
+    try {
+      parser.parseArgument(args);
+    } catch (CmdLineException e) {
+      LOG.error("Unable to parse command line arguments.", e);
+      parser.printUsage(System.err);
+      throw new IllegalArgumentException("Unable to parse command line arguments.", e);
+    }
+    return configuration;
+  }
+
+  private void startJobService() throws Exception {
+    final PrintStream oldErr = System.err;
+    ByteArrayOutputStream baos = new ByteArrayOutputStream();
+    PrintStream newErr = new PrintStream(baos);
+    try {
+      System.setErr(newErr);
+      driver =
+          FlinkJobServerDriver.fromParams(
+              new String[] {"--job-port=" + jobPort, "--artifact-port=0", "--expansion-port=0"});
+      driverThread = new Thread(driver);
+      driverThread.start();
+      boolean success = false;
+
+      // TODO: check for job service ready
+      Thread.sleep(5000);
+      success = true;
+
+      while (!success) {
+        newErr.flush();
+        String output = baos.toString(Charsets.UTF_8.name());
+        if (output.contains("JobService started on localhost:")
+            && output.contains("ArtifactStagingService started on localhost:")
+            && output.contains("ExpansionService started on localhost:")) {
+          success = true;
+        } else {
+          Thread.sleep(100);
+        }
+      }
+
+      if (!driverThread.isAlive()) {
+        throw new IllegalStateException("Job service thread is not alive");
+      }
+    } finally {
+      System.setErr(oldErr);
+    }
+  }
+
+  private void runDriverProgram() throws Exception {
+    ProcessManager processManager = ProcessManager.create();
+    String executable = "bash";
+    List<String> args =
+        ImmutableList.of("-c", String.format("exec %s " + DRIVER_CMD_FLAGS, driverCmd, jobPort));
+    String processId = "client1";
+
+    Duration timeout = Duration.ofSeconds(30);
+    Deadline deadline = Deadline.fromNow(timeout);
+    try {
+      final Process driverProcess =
+          processManager
+              .startProcess(processId, executable, args, System.getenv())
+              .getUnderlyingProcess();
```

Comment:
I think we need to extend ProcessManager instead of (a) breaking its abstraction here and (b) implementing a custom thread management on top of an existing one.